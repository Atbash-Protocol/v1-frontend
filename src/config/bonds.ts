import { Networks } from "constants/blockchain";
import { BondType } from "helpers/bond/constants";
import { BondConfig, BondProviderEnum } from "lib/bonds/bonds.types";
import { ActiveTokensEnum } from "./tokens";

import MimIcon from "assets/tokens/BASH-DAI.png";
import DAIIcon from "assets/tokens/DAI.e.png";

const BASH_DAI: BondConfig = {
    name: "bash_dai_lp",
    displayName: "BASH-DAI LP",
    token: ActiveTokensEnum.DAI,
    iconPath: MimIcon,
    lpProvider: BondProviderEnum.UNISWAP_V2, // sushi rinkeby: "https://app.sushi.com/add/0x6C538aDf35d1927497090e6971Fc46D8ed813dF6/0xdc7B08BB2AbcE1BA5b82509115F3fb7358E412aB",
    type: BondType.LP,
    addresses: {
        [Networks.MAINNET]: {
            bondAddress: "",
            reserveAddress: "",
        },
        [Networks.LOCAL]: {
            bondAddress: "0x7bc06c482DEAd17c0e297aFbC32f6e63d3846650",
            reserveAddress: "0x908B40ED87FCA620B101B3Cba2B4C640D11eF016",
        },
        [Networks.RINKEBY]: {
            bondAddress: "0xcE24D6A45D5c59D31D05c8C278cA3455dD6a43DA",
            reserveAddress: "0x26DF06b47412dA76061ddA1fD9fe688A497FB88b", // uniswapv2-"0xC35F84DBd48fcB0467ac3Ee2C4e37D848B8d3173",
        },
    },
    isActive: false,
};

const StableDAI: BondConfig = {
    name: "DAI",
    displayName: "DAI",
    token: ActiveTokensEnum.DAI,
    iconPath: DAIIcon,
    lpProvider: BondProviderEnum.UNISWAP_V2, // sushi rinkeby: "https://app.sushi.com/add/0x6C538aDf35d1927497090e6971Fc46D8ed813dF6/0xdc7B08BB2AbcE1BA5b82509115F3fb7358E412aB",
    type: BondType.StableAsset,
    addresses: {
        [Networks.RINKEBY]: {
            bondAddress: "0xd0D5024c723c4F047C7626FBA18E7797A4E50a12",
            reserveAddress: "0xdc7B08BB2AbcE1BA5b82509115F3fb7358E412aB", // DAI
        },
        [Networks.LOCAL]: {
            bondAddress: "0x4ed7c70F96B99c776995fB64377f0d4aB3B0e1C1",
            reserveAddress: "0xDc64a140Aa3E981100a9becA4E685f962f0cF6C9", // DAI
        },
        [Networks.MAINNET]: {
            bondAddress: "",
            reserveAddress: "",
        },
    },
    isActive: true,
};

export const BONDS = [BASH_DAI, StableDAI];
