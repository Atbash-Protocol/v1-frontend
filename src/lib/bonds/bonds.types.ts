import { Networks } from 'constants/blockchain';

import { BondOptions } from './bond/bond';

export enum BondType {
    STABLE_ASSET,
    LP,
    CUSTOM,
}

export enum BondProviderEnum {
    UNISWAP_V2 = 'https://app.uniswap.org/#/add/v2',
}

export interface BondAddresses {
    reserveAddress: string;
    bondAddress: string;
}

export interface NetworkAddresses {
    [Networks.MAINNET]?: BondAddresses;
    [Networks.RINKEBY]: BondAddresses;
    [Networks.LOCAL]: BondAddresses;
}

export type BondConfig = Omit<BondOptions, 'networkID'> & {
    addresses: NetworkAddresses;
    isActive?: boolean;
};
