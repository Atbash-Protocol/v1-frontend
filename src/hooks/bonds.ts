import { useMemo } from 'react';

import { useSelector } from 'react-redux';

import { BondItem } from 'store/modules/bonds/bonds.types';
import { IReduxState } from 'store/slices/state.interface';

// Slaps together bond data within the account & bonding states
const useBonds = () => {
    const bonds = useSelector<IReduxState, BondItem[]>(state => Object.values(state.bonds.bonds));

    const mostProfitableBonds = useMemo(
        () =>
            bonds.sort((bond1, bond2): number => {
                if (bond1.metrics.bondDiscount === null || bond2.metrics.bondDiscount === null) return 0;

                return bond1.metrics.bondDiscount > bond2.metrics.bondDiscount ? 0 : 1;
            }),
        [bonds],
    );

    return {
        bonds,
        mostProfitableBonds,
    };
};

export default useBonds;

export const selectBondReady = (bond: BondItem) => {
    return Object.values(bond.metrics).some(metric => metric !== null);
};

export const useBondPurchaseReady = () => {
    const metrics = useSelector<IReduxState, boolean>(state => state.main.metrics.reserves !== null);

    const marketReady = useSelector<IReduxState, boolean>(state => state.markets.markets.dai !== null);

    return metrics && marketReady;
};
