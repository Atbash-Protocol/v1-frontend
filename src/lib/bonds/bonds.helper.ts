import { Networks } from 'constants/blockchain';

import { BondOptions } from './bond/bond';
import { LPBond } from './bond/lp-bond';
import { StableBond } from './bond/stable-bond';
import { BondAddresses, BondConfig, BondType } from './bonds.types';

export const createBond = (bondConfig: BondOptions): LPBond | StableBond => {
    switch (bondConfig.type) {
        case BondType.LP:
            return new LPBond(bondConfig);
        case BondType.StableAsset:
            return new StableBond(bondConfig);

        default:
            throw new Error('Unedefined bond type');
    }
};

export const getBondContracts = (bondConfig: BondConfig, networkID: Networks): BondAddresses => {
    const addresses = bondConfig.addresses[networkID];

    if (!addresses || !addresses.bondAddress || !addresses.reserveAddress) throw new Error('Unable to get bond addresses');

    return { bondAddress: addresses.bondAddress, reserveAddress: addresses.reserveAddress };
};
