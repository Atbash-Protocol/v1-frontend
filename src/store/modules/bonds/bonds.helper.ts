import { BondMetrics } from "./bonds.types";

export const initDefaultBondMetrics = (): BondMetrics => ({
    treasuryBalance: null,
    bondDiscount: null,
    bondQuote: null,
    purchased: null,
    vestingTerm: null,
    maxBondPrice: null,
    bondPrice: null,
    marketPrice: null,
    maxBondPriceToken: null,
});
