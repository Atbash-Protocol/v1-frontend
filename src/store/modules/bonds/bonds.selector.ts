import { createDraftSafeSelector } from '@reduxjs/toolkit';
import Decimal from 'decimal.js';
import { TFunction } from 'i18next';
import { isNil, sum } from 'lodash';
import { Duration } from 'luxon';
import { createSelector } from 'reselect';

import { formatTimer } from 'helpers/prettify-seconds';
import { formatUSD, formatUSDFromDecimal } from 'helpers/price-units';
import { LPBond } from 'lib/bonds/bond/lp-bond';
import { StableBond } from 'lib/bonds/bond/stable-bond';
import { RootState } from 'store/store';

import { useContractLoaded } from '../app/app.selectors';
import { selectIndex } from '../stake/stake.selectors';
import { BondItem, BondMetrics } from './bonds.types';

export const selectBondInstances = (state: RootState) => Object.values(state.bonds.bondInstances);

export const selectAllBonds = createDraftSafeSelector([selectBondInstances], bondInstances => {
    return Object.values(bondInstances).reduce(
        (acc, bondInstance) => {
            if (bondInstance.bondOptions.isActive === true) {
                return {
                    ...acc,
                    activeBonds: [...acc.activeBonds, bondInstance],
                };
            }

            return {
                ...acc,
                inactiveBonds: [...acc.inactiveBonds, bondInstance],
            };
        },
        { activeBonds: new Array<LPBond | StableBond>(), inactiveBonds: new Array<LPBond | StableBond>() },
    );
});

export const selectBondInfos = (bonds: Record<string, BondItem>, bondID: string): BondItem | null => {
    const bond = bonds[bondID];

    if (!bond) return null;

    return bond;
};

export const selectBondMintingMetrics = (metrics: BondMetrics) => {
    let bondPrice = null;

    bondPrice = formatUSD(Number(metrics.bondPrice) / 1e18, 2);

    const balance = metrics.balance ? new Decimal(metrics.balance.toString()).div(10 ** 18).toFixed(2) : null;
    const quote = !isNil(metrics.bondQuote) ? ` ${new Decimal(metrics.bondQuote).toFixed(2)} BASH` : null;
    const vestingTerm = metrics.vestingTerm ? Duration.fromObject({ seconds: metrics.vestingTerm }).as('days') : null;

    return {
        bondPrice,
        allowance: metrics.allowance,
        balance,
        vestingTerm,
        maxBondPrice: metrics.maxBondPrice,
        bondDiscount: !isNil(metrics.bondDiscount) ? `${metrics.bondDiscount.mul(100).toFixed(2)} %` : null,
        purchased: metrics.purchased !== null ? formatUSD(metrics.purchased) : null,
        bondSoldOut: (metrics.bondDiscount?.toNumber() ?? 0) * 100 < -30,
        quote,
    };
};

export const selectBondPrice = (state: RootState, bondID: string) => {
    const metrics = state.bonds.bondMetrics[bondID];

    if (!metrics || metrics.bondPrice === null) return null;

    return formatUSDFromDecimal(new Decimal(metrics.bondPrice.toString()).div(10 ** 18), 2);
};

export const selectBondIsQuoting = (state: RootState, bondID: string) => {
    const bond = state.bonds.bondMetrics[bondID];

    if (!bond) throw new Error('Unable to get bond');

    return bond.loading ?? false;
};

export const selectBondQuoteResult = (
    {
        bonds: { bondQuote },
        main: {
            blockchain: { timestamp },
        },
    }: RootState,
    t: TFunction,
) => {
    return {
        interestDue: bondQuote.interestDue,
        vesting: timestamp === null || bondQuote.bondMaturationBlock === null ? null : formatTimer(timestamp ?? 0, bondQuote.bondMaturationBlock ?? 0, t),
        pendingPayout: bondQuote.pendingPayout,
    };
};

export const selectBondMetrics = (state: RootState) => Object.values(state.bonds.bondMetrics);

export const selectBondItemMetrics = (state: RootState, bondID: string) => state.bonds.bondMetrics[bondID];

export const selectMostProfitableBonds = createSelector([selectBondMetrics], bonds => {
    const orderedBonds = bonds.sort((bond1, bond2): number => {
        if (bond1.bondDiscount === null || bond2.bondDiscount === null) return 0;

        return bond1.bondDiscount > bond2.bondDiscount ? 0 : 1;
    });

    return orderedBonds;
});

export const selectBondsReady = createSelector([selectBondMetrics, useContractLoaded], (bonds, contractLoaded) => {
    return bonds.length > 0 && contractLoaded;
});

// refactor

export const selectBondInstance = (state: RootState, bondId: string) => {
    const bond = state.bonds.bondInstances[bondId];

    if (!bond) return null;

    return bond;
};

export const selectBondMetricsReady = (state: RootState, bondId: string) => {
    const bondMetrics = state.bonds.bondMetrics[bondId];

    if (!bondMetrics) return false;

    return Object.values(bondMetrics).some(v => v !== null);
};

export const selectAllBondMetrics = (state: RootState) => state.bonds.bondMetrics;

export const selectBondDetailsCalcReady = createSelector([selectAllBondMetrics], bondMetrics => {
    const ready = Object.values(bondMetrics).some(v => v.bondPrice !== null);

    return ready;
});

export const selectFormattedTreasuryBalance = createSelector([selectAllBondMetrics], metrics => {
    const total = sum(
        Object.values(metrics)
            .map(m => m.treasuryBalance)
            .filter(Number),
    );

    return formatUSD(total, 2);
});

export const selectTreasuryReady = createSelector([selectAllBondMetrics], bondMetrics => {
    return Object.entries(bondMetrics)
        .map(([, bond]) => bond.treasuryBalance)
        .some(balance => balance !== null);
});

const selectCoreMetrics = (state: RootState) => state.bonds.coreMetrics;

export const selectFormattedBondCoreMetrics = createSelector([selectCoreMetrics, selectIndex], ({ rfv, runway }, currentIndex) => {
    const indexInGwei = new Decimal((currentIndex ?? 0).toString()).div(10 ** 9);
    const rfvBASH = new Decimal(rfv).mul(indexInGwei).toNumber();

    return {
        rfv: formatUSD(rfv),
        rfvBASH: formatUSD(rfvBASH),
        runway: new Decimal(runway).toFixed(1),
    };
});

export const selectAllActiveBondsIds = createSelector([selectAllBonds], ({ activeBonds }) => Object.values(activeBonds).map(({ ID }) => ID));
