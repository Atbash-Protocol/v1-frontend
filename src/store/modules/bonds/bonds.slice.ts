import { createSlice } from '@reduxjs/toolkit';

import {
    approveBonds,
    calcBondDetails,
    calculateUserBondDetails,
    getBondTerms,
    initializeBonds,
    loadBondBalancesAndAllowances,
    getTreasuryBalance,
    getBondMetrics,
} from './bonds.thunks';
import { BondSlice } from './bonds.types';

// Define the initial state using that type
const initialState: BondSlice = {
    bonds: {},
    bondInstances: {},
    bondMetrics: {},
    coreMetrics: {
        // common metrics for all bonds
        rfv: 0,
        rfvTreasury: 0,
        runway: 0,
        deltaMarketPriceRfv: 0,
    },
    bondCalculator: null,
    treasuryBalance: null,
    loading: true,
    bondQuoting: false,
    bondQuote: {
        interestDue: null,
        bondMaturationBlock: null,
        pendingPayout: null,
    },
};

export const BondSlices = createSlice({
    name: 'app-bonds',
    initialState,
    reducers: {},
    extraReducers: builder => {
        builder.addCase(initializeBonds.fulfilled, (state, { payload: { bondCalculator, bondMetrics, bondInstances } }) => {
            return {
                ...state,
                // eslint-disable-next-line @typescript-eslint/no-explicit-any
                bondCalculator: bondCalculator as any,
                bondInstances,
                bondMetrics,
            };
        });

        builder.addCase(getTreasuryBalance.fulfilled, (state, action) => {
            for (const [bondID, treasury] of Object.entries(action.payload)) {
                state.bondMetrics[bondID].treasuryBalance = treasury as number;
            }
        });

        builder.addCase(
            calcBondDetails.pending,
            (
                state,
                {
                    meta: {
                        arg: { bondID },
                    },
                },
            ) => {
                state.bondMetrics[bondID].loading = true;
            },
        );

        builder.addCase(calcBondDetails.fulfilled, (state, { payload: { bondID, ...metrics } }) => {
            return {
                ...state,
                bondMetrics: {
                    ...state.bondMetrics,
                    [bondID]: {
                        ...state.bondMetrics[bondID],
                        ...metrics,
                        loading: false,
                    },
                },
            };
        });

        builder.addCase(getBondTerms.fulfilled, (state, { payload, meta: { arg: bondID } }) => {
            state.bondMetrics[bondID].terms = payload.terms;
        });

        builder.addCase(approveBonds.fulfilled, (state, { payload, meta }) => {
            state.bondMetrics[meta.arg.bondID].allowance = payload?.allowance ?? null;
        });

        builder.addCase(
            loadBondBalancesAndAllowances.fulfilled,
            (
                state,
                {
                    payload,
                    meta: {
                        arg: { bondID },
                    },
                },
            ) => {
                state.bondMetrics[bondID].allowance = payload.allowance;
                state.bondMetrics[bondID].balance = payload.balance;
            },
        );

        builder.addCase(calculateUserBondDetails.pending, (state, {}) => {
            state.bondQuote = {
                interestDue: null,
                bondMaturationBlock: null,
                pendingPayout: null,
            };
            state.bondQuoting = true;
        });
        builder.addCase(calculateUserBondDetails.fulfilled, (state, { payload }) => {
            state.bondQuote = payload;
            state.bondQuoting = false;
        });

        builder.addCase(getBondMetrics.fulfilled, (state, { payload }) => {
            state.coreMetrics.rfv = payload.rfv;
            state.coreMetrics.rfvTreasury = payload.rfvTreasury;
            state.coreMetrics.runway = payload.runway;
            state.coreMetrics.deltaMarketPriceRfv = payload.deltaMarketPriceRfv;
        });
    },
});

export default BondSlices.reducer;
