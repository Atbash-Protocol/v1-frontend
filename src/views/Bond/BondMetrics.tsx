import { Grid } from '@mui/material';
import { useTranslation } from 'react-i18next';

import MemoInlineMetric from 'components/Metrics/InlineMetric';
import { selectBondMintingMetrics } from 'store/modules/bonds/bonds.selector';
import { BondMetrics } from 'store/modules/bonds/bonds.types';

const CBondMetrics = ({ bondMetrics }: { bondMetrics: BondMetrics }) => {
    const { t } = useTranslation();

    const metrics = selectBondMintingMetrics(bondMetrics);

    if (!metrics) {
        return <> </>;
    }

    const { maxBondPrice, bondDiscount, vestingTerm } = metrics;

    const metrics2 = [
        { value: (maxBondPrice ?? '').toString(), metricKey: t('bond:MaxYouCanBuy') },
        {
            metricKey: t('bond:ROI'),
            value: bondDiscount,
        },
        { metricKey: t('bond:MinimumPurchase'), value: vestingTerm },
    ].map(({ value, metricKey }, index) => <MemoInlineMetric {...{ value, metricKey }} key={index} />);

    return (
        <Grid xs={12} container>
            {metrics2}
        </Grid>
    );
};

export default CBondMetrics;
