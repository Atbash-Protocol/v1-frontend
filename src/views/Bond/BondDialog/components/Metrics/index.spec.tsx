import { render } from '@testing-library/react';

import * as BondSelectorModule from 'store/modules/bonds/bonds.selector';
import { renderComponent } from 'tests/utils';

import BondMetrics from '.';

describe('BondMetrics', () => {
    it('renders the metrics', () => {
        jest.spyOn(BondSelectorModule, 'selectBondMintingMetrics').mockReturnValue({
            maxBondPrice: 10,
            balance: '20 BASH',
            quote: '$2000',
            bondDiscount: '50%',
            vestingTerm: 10,
        } as any);

        const bondMetrics = {
            treasuryBalance: null,
            bondDiscount: null,
            bondQuote: null,
            purchased: null,
            vestingTerm: null,
            maxBondPrice: null,
            bondPrice: null,
            marketPrice: null,
            maxBondPriceToken: null,
            allowance: null,
            balance: null,
        };
        const { container } = render(<BondMetrics bondMetrics={bondMetrics} />);

        expect(container).toMatchSnapshot();
    });
});
