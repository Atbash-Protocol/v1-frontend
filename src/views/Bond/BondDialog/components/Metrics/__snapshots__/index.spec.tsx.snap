// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BondMetrics renders the metrics 1`] = `
<div>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-grid-xs-12 css-spompt-MuiGrid-root"
  >
    <div
      class="MuiBox-root css-1z137pa"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 css-1qha2kt-MuiTypography-root"
      >
        bond:YourBalance
      </p>
      <p
        class="MuiTypography-root MuiTypography-body1 css-le79b4-MuiTypography-root"
      >
        20 BASH
      </p>
    </div>
    <div
      class="MuiBox-root css-1z137pa"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 css-1qha2kt-MuiTypography-root"
      >
        bond:YouWillGet
      </p>
      <p
        class="MuiTypography-root MuiTypography-body1 css-le79b4-MuiTypography-root"
      >
        $2000
      </p>
    </div>
    <div
      class="MuiBox-root css-1z137pa"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 css-1qha2kt-MuiTypography-root"
      >
        bond:MaxYouCanBuy
      </p>
      <p
        class="MuiTypography-root MuiTypography-body1 css-le79b4-MuiTypography-root"
      >
        $10.00
      </p>
    </div>
    <div
      class="MuiBox-root css-1z137pa"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 css-1qha2kt-MuiTypography-root"
      >
        bond:Discount
      </p>
      <p
        class="MuiTypography-root MuiTypography-body1 css-le79b4-MuiTypography-root"
      >
        50%
      </p>
    </div>
    <div
      class="MuiBox-root css-1z137pa"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 css-1qha2kt-MuiTypography-root"
      >
        bond:VestingTerm
      </p>
      <p
        class="MuiTypography-root MuiTypography-body1 css-le79b4-MuiTypography-root"
      >
        common:day
      </p>
    </div>
    <div
      class="MuiBox-root css-1z137pa"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 css-1qha2kt-MuiTypography-root"
      >
        bond:MinimumPurchase
      </p>
      <p
        class="MuiTypography-root MuiTypography-body1 css-le79b4-MuiTypography-root"
      >
        0.01 BASH
      </p>
    </div>
  </div>
</div>
`;
